compiler/hash-layout/src/ty.rs

```diff
@@ -0,0 +1,1136 @@
+    }
+
+    /// Check if the type is a raw `str` type.
+    pub fn is_str(&self) -> bool {
+        self.borrow().is_str()
+    }
+}
+
+static_sequence_store_indirect!(
+    store = pub ReprTyListStore,
+    id = pub ReprTyListId[ReprTyId],
+    store_name = ty_list,
+    store_source = repr_stores()
+);
+
+impl ReprTyListId {
+    pub fn seq<I: IntoIterator<Item = ReprTyId>>(values: I) -> Self {
+        repr_stores().ty_list().create_from_iter(values)
+    }
+}
+
+/// Macro that is used to create the "common" IR types. Each
+/// entry has an associated name, and then followed by the type
+/// expression that represents the [ReprTy].
+macro_rules! create_common_ty_table {
+    ($($name:ident: $value:expr),* $(,)?) => {
+
+        /// Defines a map of common types that might be used in the IR
+        /// and general IR operations. When creating new types that refer
+        /// to these common types, they should be created using the
+        /// using the associated [ReprTyId]s of this map.
+        pub struct CommonReprTys {
+            $(pub $name: ReprTyId, )*
+        }
+
+        impl CommonReprTys {
+            pub fn new() -> CommonReprTys {
+                // Create a `unit` type in order to reserve the first index of
+                // the ADT for a `()` type.
+                let _ = ReprTy::tuple(&[]);
+                $(let $name = ReprTy::create($value); )*
+
+                CommonReprTys {
+                    $($name,)*
+                }
+            }
+        }
+
+        impl Default for CommonReprTys {
+            fn default() -> Self {
+                Self::new()
+            }
+        }
+    };
+}
+
+create_common_ty_table!(
+    // ------------------------------------------
+    // Primitive types
+    // ------------------------------------------
+    bool: ReprTy::Bool,
+    char: ReprTy::Char,
+    never: ReprTy::Never,
+
+    // ------------------------------------------
+    // Floating point types
+    // ------------------------------------------
+    f32: ReprTy::Float(FloatTy::F32),
+    f64: ReprTy::Float(FloatTy::F64),
+
+    // ------------------------------------------
+    // Signed integer types
+    // ------------------------------------------
+    i8: ReprTy::Int(SIntTy::I8),
+    i16: ReprTy::Int(SIntTy::I16),
+    i32: ReprTy::Int(SIntTy::I32),
+    i64: ReprTy::Int(SIntTy::I64),
+    i128: ReprTy::Int(SIntTy::I128),
+    isize: ReprTy::Int(SIntTy::ISize),
+
+    // ------------------------------------------
+    // Unsigned integer types
+    // ------------------------------------------
+    u8: ReprTy::UInt(UIntTy::U8),
+    u16: ReprTy::UInt(UIntTy::U16),
+    u32: ReprTy::UInt(UIntTy::U32),
+    u64: ReprTy::UInt(UIntTy::U64),
+    u128: ReprTy::UInt(UIntTy::U128),
+
+    // ------------------------------------------
+    // Unit types, and unit ptr types
+    // ------------------------------------------
+    usize: ReprTy::UInt(UIntTy::USize),
+    unit: ReprTy::Adt(AdtId::UNIT),
+
```

Comment:
What is the point of having all these special variants in `ReprTy` if they are created as instances of `ReprTy` here anyway? Why not reduce the variants of `ReprTy` to just N * some number of bytes instead, and these common types can remain as they are now?

Reply:
Perhaps I am not realising something

Reply:
Well, the `Bytes` idea isn't so bad to add to the variants, however we will still need things like a machine specific size variant.

Reply:
Also, `bool` and `u8` aren't gonna be treated in the same manner, and hence we'll need a way to distinguish them.

I think we could for the most part remove the integer types and add the following variant:

```rs

enum ReprTy {
  Bool,

  /// Floats are always specific sizes anyway
  Float(..),


  /// chars can also be put here
  Bytes {
    signed: bool,
    size: usize
  },

  MachineInt { 
    signed: bool
  },

  ..
}
```

So yeah this is an option, it is probably better than the current and allows for more flexibility in the future.

Reply:
I also think right now is probably not the right time to make this representation change. Perhaps when we finish the Const in TIR refactor and the various changes that you are making, then we can consider a representational change.

Reply:
I am curious what the difference in handling between bool and u8 is